include/ruby/internal/intern/hash.h

```diff
@@ -106,6 +106,16 @@ VALUE rb_hash(VALUE obj);
  */
 VALUE rb_hash_new(void);
 
+/**
+ * Identical to rb_hash_new(), except it additionally specifies how many keys
+ * it is expected to contain. This way you can create a hash that is large enough
+ * for your need and won't need to be reallocated and rehashed multiple times.
```

Comment:
[Doc] "won't be reallocated" could be too strong.  Keys can collide.  It could be possible for rehash to happen when multiple keys have identical hash value.  We don't do that today, but we did before.  This is rather an implementation detail than a spec.  Alternative implementations could choose otherwise.

Reply:
Thanks, I tried to rephrase:

> This way you can create a hash that is large enough for your need. For large hashes it means it won't need to be reallocated and rehashed as much, improving performance.